File: coindesk/src/app/service/currency.service.ts
import { Injectable } from '@angular/core';
import { HttpClient, HttpParams } from '@angular/common/http';
import { Observable } from 'rxjs';
import { Currency } from '../model/currency.interface';
import { TransformedCurrency } from './exchange-rate.service';

@Injectable({
  providedIn: 'root',
})
export class CurrencyService {
  private apiUrl = 'http://localhost:8080/api/currency';

  constructor(private http: HttpClient) {}

  getAllCurrencies(): Observable<Currency[]> {
    return this.http.get<Currency[]>(this.apiUrl);
  }

  createOrUpdate(currency: TransformedCurrency): Observable<void> {
    return this.http.post<void>(`${this.apiUrl}/createOrUpdate`, currency);
  }

  deleteCurrency(code: string): Observable<void> {
    const params = new HttpParams().set('code', code);
    return this.http.post<void>(`${this.apiUrl}/delete`, null, { params });
  }
}

File: coindesk-api/src/main/java/com/example/coindesk_api/Service/CurrencyService.java
package com.example.coindesk_api.Service;
import java.util.List;

import com.example.coindesk_api.DTO.TransformedCurrency;
import com.example.coindesk_api.Domain.CurrencyEntity;

public interface CurrencyService {
    public List<CurrencyEntity> getAllCurrencies();
    public void createOrUpdate(TransformedCurrency transformedCurrency);
    public void deleteCurrency(String code);
}
